MailManager/src/tests/java/inc/softserve/EnvelopeServiceTest.java

```diff
@@ -0,0 +1,78 @@
+package inc.softserve;
+
+import inc.softserve.common.EnvelopeTools;
+import inc.softserve.dao.EnvelopeDao;
+import inc.softserve.dao.UserDao;
+import inc.softserve.services.EnvelopeService;
+import inc.softserve.services.EnvelopeServiceImpl;
+import org.testng.Assert;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class EnvelopeServiceTest extends Assert {
+
+    EnvelopeService service;
+
+    @BeforeClass
+    public void setUpDataSource() throws Exception {
+
+        UserDao usrDao = mock(UserDao.class);
+        User user = new User("akonon", "a", "b");
+        when(usrDao.getUserByUsername(anyString())).thenReturn(null);
+        when(usrDao.getUserByUsername("akonon")).thenReturn(user);
+
+        Envelope envelopeNotFound = new Envelope(user, "akonon22@mail.ru", "TestEmail", "Some template", EnvelopeState.NOT_FOUND);
+        Envelope envelopeSent = new Envelope(user, "akonon22@mail.ru", "TestEmail", "Some template", EnvelopeState.SENT);
+        EnvelopeDao envDao = mock(EnvelopeDao.class);
+        when(envDao.saveOrUpdate(any(Envelope.class))).thenAnswer((invocation) -> {Object[] args = invocation.getArguments();
+                                                                                    return (Envelope) args[0];});
+        when(envDao.getById("goodId")).thenReturn(envelopeSent);
+        when(envDao.getById("badId")).thenReturn(envelopeNotFound);
+        when(envDao.getById("illegalId")).thenReturn(null);
+
+        EnvelopeTools envTools = mock(EnvelopeTools.class);
+        when(envTools.sendEnvelope(any(Envelope.class))).thenReturn(EnvelopeState.SENT)
+                                                        .thenReturn(EnvelopeState.IN_PROGRESS)
+                                                        .thenReturn(EnvelopeState.ERROR);
+
+        service = new EnvelopeServiceImpl(envDao, usrDao, envTools);
+    }
+
+    @DataProvider(name = "saveOrUpdateData")
+    public Object[][] saveOrUpdateData() {
+        return new Object[][]{
+                {"TestEmail", "akonon22@mail.ru", "akonon", "Some template", EnvelopeState.SENT},
+                {"TestEmail", "akonon22@mail.ru", "akonon", "Some template", EnvelopeState.IN_PROGRESS},
+                {"TestEmail", "akonon22@mail.ru", "akonon", "Some template", EnvelopeState.ERROR},
```

Comment:
That may have sense to extract those string constants either in local method variables, or in static fields.